activerecord/lib/active_record/associations/association.rb

```diff
@@ -377,10 +377,16 @@ def inversable?(record)
 
         def matches_foreign_key?(record)
           if foreign_key_for?(record)
-            record.read_attribute(reflection.foreign_key) == owner.id ||
-              (foreign_key_for?(owner) && owner.read_attribute(reflection.foreign_key) == record.id)
+            record.read_attribute(reflection.foreign_key) == owner.read_attribute(reflection.association_primary_key) ||
```

Comment:
> I think we can never get to the second part of this check

what second part do you mean? 

I am little confused by your example given that 

```
(foreign_key_for?(record) && record.read_attribute(reflection.foreign_key) == owner.read_attribute(reflection.association_primary_key))
```

is the same as 
```
if foreign_key_for?(record)
   record.read_attribute(reflection.foreign_key) == owner.read_attribute(reflection.association_primary_key)
```

Reply:
Thats not necessarily true. The test added in this change would demonstrate where the association is self referencing which in this case it will have the same foreign key for both the record and the owner.

Reply:
Agreed, thanks for pointing that out. You're right that both the models can have the foreign_key in the self referencing case. I guess what I meant is that the foreign_key is always on either side of the _association_, semantically. So in a HasOne/HasMany association, the foreign key is on the other side of the association, and for BelongsTo, its on the 'owner' side of the association. If my mental model is right, reading the association's foreign_key attribute on the owner in a HasOne/Many association doesn't make sense, even if it exists.

The code checks that foreign_key_for?(record), and if its true, we handle the case that this is either a HasOne/Many, or BelongsTo association, as we check both record.foreign_key == owner.primary_key and vice versa. However, I believe if this association is a polymorphic BelongsTo, we run into the same issue we are guarding against in the else block. 

If foreign_key_for?(record is false), then I think the association _must_ be a BelongsTo association, owning to the 'either side of the association' thing mentioned above. Also, `polymorphic?` only has [meaning](https://github.com/rails/rails/blob/744b6715a5d07a2ddffceea8283f7b787541bc60/activerecord/lib/active_record/reflection.rb#L634) on a BelongsTo association(its always false on a HasMany/HasOne). And in the polymorphic case, it seems like we shouldn't be ignoring the passed in custom primary key!

I hope that makes sense, and I appreciate your patient responses. I'm freshly looking at this section of rails and so perhaps lots of these comments are due to my relative lack of familiarity. Im happy to further comment more if you this makes sense, otherwise feel free to ignore - I roughly have an idea on how this can be improved, but ill need to create some test cases  to validate my thinking.